### PR TITLE
[codegen/docs] Generate a package tree that can be serialized as JSON

### PR DESCRIPTION
**Pull request by praneetloke**
# Description

This PR adds a new exported function in the docs generator that can be used by the `resourcedocsgen` tool to retrieve a package tree that can be emitted somewhere for our docs site. The JSON file is used to render some UI components.

Along with this, I have also updated the docs template bundler to be invoked programmatically. It's useful for the SDK generator tests we run from `docs/gen_test.go` so that it doesn't end up using a stale `packaged.go` which is git ignored.

Fixes https://github.com/pulumi/pulumi-service/issues/7346

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change
  - **Note**: I didn't do this because this has no end user impact. Let me know if you still would like me to include this in the pending list
- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/codegen/docs/bundler/bundler.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
+package bundler
```
Comment (praneetloke) [inline pkg/codegen/docs/bundler/bundler.go:1]: All of this was moved from `docs/bundler.go`.

File: pkg/codegen/docs/package_tree.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+// navigable tree format starting from the top-level/index/root
+// of a package.
+type PackageTreeItem struct {
+	Name     string            `json:"name"`
+	Type     entryType         `json:"type"`
+	Link     string            `json:"link"`
+	Children []PackageTreeItem `json:"children,omitempty"`
+}
+
+func generatePackageTree(rootMod modContext) ([]PackageTreeItem, error) {
```
Comment (praneetloke) [inline pkg/codegen/docs/package_tree.go:25]: I could also make this a receiver function of the `modContext` type if you think that reads better. I started going down this route and didn't think it was bad enough to have to change to that.

File: pkg/codegen/docs/gen.go [hunk 3]
```diff
@@ -1831,9 +1827,7 @@ func generateModulesFromSchemaPackage(tool string, pkg *schema.Package) map[stri
 	return modules
 }
 
-// GeneratePackage generates the docs package with docs for each resource given the Pulumi
-// schema.
-func GeneratePackage(tool string, pkg *schema.Package) (map[string][]byte, error) {
+func Initialize(tool string, pkg *schema.Package) {
```
Comment (praneetloke) [inline pkg/codegen/docs/gen.go:1830]: Adding this initialize allows to use the generated modules to do other things with it as separate operations rather a single function. It seemed like a good idea to decompose them into separate things.

**Review by justinvp — APPROVED** [review]

**Comment by praneetloke** [comment]
The downstream resource docs workflow failure is expected because of some changes. The `resourcedocsgen` tool in `docs` repo uses a local path override for `pulumi/pulumi`. So there will be a dependency problem until the associated PR is https://github.com/pulumi/docs/pull/6530 merged, which is also the PR that actually uses these changes. So yeah these are inter-dependent right now in their PR workflows.

As for https://github.com/pulumi/pulumi/issues/7976, it seems to be a flaky test. See https://github.com/pulumi/pulumi/issues/7976.

**Comment by praneetloke** [comment]
Alright the only failing PR workflow is the resource docs test. As mentioned before, that's ok to ignore for this PR. Note that all SDK gen unit tests that include docs generation passed. I'll merge this.

**Pull request merged**